### PR TITLE
Add V3 SpeedDistance and Location graphs

### PR DESCRIPTION
**Pull request by haotongwang**
## Description

Adds the SpeedDistanceChart and LocationMap (renamed from LocationTimeChart) components for V3.
Also includes a skeleton of the dashboard view.

## Screenshots

image

## Steps to Test

1. Start `client` and `server` and `mosquitto` according to dashboard repo instructions

2. Execute the V3 fake module DAS from [DAS #71](https://github.com/monash-human-power/data-acquisition-system/pull/73)

### CODE REVIEW

**Comment by haotongwang** [comment]
Just want to say, it looks like there's a lot in this PR but most of it is fixing imports and types in a bunch of files.

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (haotongwang) [inline client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx]: Do we want to implement a multi sensor data hook? Currently, the usage is finding the correct sensor from the notion mqtt topics page.
Not sure if that's something we want in the MVP or not.

#108 also has the same problem.

File: client/src/components/common/charts/SpeedDistanceChart.tsx
Comment (kdav108) [inline client/src/components/common/charts/SpeedDistanceChart.tsx:10]: I'm not sure if this is a convention or something, but would it be possible to make the parameters more detailed (i.e explain what `data` and `max` is)

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (kdav108) [inline client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx:62]: Probably the last few times we can say 'You forgot to re-format with prettier' before the new 'prettier on commit' PR merges in.

File: client/src/components/common/charts/SpeedDistanceChart.tsx
Comment (kdav108) [inline client/src/components/common/charts/SpeedDistanceChart.tsx]: A bit pedantic but can we have a more conventional title like 'Speed vs Distance'

**Review by kdav108 — DISMISSED** [review]
I ran the entire suite and from the look and functionality of the graph and the map it feels really complicated to build (at least for me 😂). So really great job with this feature!

Everything looks good nothing too significant apart from the following UI concerns:
- On my chrome browser the graph and the map don't appear side-by-side as intended unless I zoom out to 90%, is this expected?
- When I stop and re-run `V3_fake_module.py`, the speed vs distance graph starts drawing from the beginning (left side of the graph), but still maintains the previous graph (created when I first ran the fake module script). See the image below:
Screen Shot 2021-03-09 at 9 29 22 pm
I would assume we might want the graph to start from where it left off or clear the graph entirely and restart sketching in the event that the Wireless modules stop and restart.

File: client/src/components/common/charts/SpeedDistanceChart.tsx
Comment (haotongwang) [reply client/src/components/common/charts/SpeedDistanceChart.tsx:10]: The documentation for `props: ChartProps` can be found in `types/chart` so has been omitted here.

File: client/src/components/common/charts/SpeedDistanceChart.tsx
Comment (haotongwang) [reply client/src/components/common/charts/SpeedDistanceChart.tsx]: The convention from V2 is 'X-Y' so I've continued with that. It also makes it a bit easier to say. Happy to change if there's a consensus.

**Comment by haotongwang** [comment]
@kdav108 The graphs seems to work fine for me on 100%, chrome. What resolution are you working with? 
image

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (haotongwang) [reply client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx:62]: Not sure what this is referring to, the 4 space indentation?

**Comment by haotongwang** [comment]
I'm not too sure about how exactly we want the graphs to interact with the data and when it should reset. For example, the DAS could disconnect during a run, and in that case, we would still want to preserve the previous data. Maybe @hallgchris can shed some light on this?

File: client/.vscode/settings.json
Comment (hallgchris) [inline client/.vscode/settings.json]: Do people usually have `client`/`server` as the project route, or the repository root? I use the latter, and it seems vscode doesn't pick this up.

I also wonder if we should be committing stuff like this in the first place, as it'll lead to conflicts if people want their own personal vscode settings.

File: client/src/components/common/charts/CadenceTimeChart.js
Comment (hallgchris) [inline client/src/components/common/charts/CadenceTimeChart.js:2]: If it's possible, can you see if your import formatter to be configured to put library imports above our own? It's the I believe we've been following across all our repos and I think it's nicer starting with the broad imports and going to specific ones. Would also mean a lot less files are touched in this PR!

File: client/src/components/common/charts/LocationMap.tsx
Comment (hallgchris) [inline client/src/components/common/charts/LocationMap.tsx]: Now that we're using typescript for the map, we can remove `PropTypes` entirely from the file!

File: client/src/components/common/charts/SpeedDistanceChart.tsx
Comment (hallgchris) [reply client/src/components/common/charts/SpeedDistanceChart.tsx]: I agree 'Speed vs Distance' would look nicer. No harm in changing it I reckon!

File: client/src/components/v3/dashboard/V3LocationMap.tsx
Comment (hallgchris) [inline client/src/components/v3/dashboard/V3LocationMap.tsx]: Latitude should range from -90 to 90

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (hallgchris) [inline client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx]: `Math.max(...data)` would be a lot more readable.

Even better, you could keep track of the maximum as each data point is added, so that it's an O(1) not O(n) operation. Probably not a big deal in practice though.

File: client/.eslintrc
Comment (hallgchris) [inline client/.eslintrc]: This change makes building storybook from scratch fail for me. Anyone else able to reproduce?
```bash
rm -r node_modules
yarn
yarn run storybook
```

**Review by hallgchris — CHANGES_REQUESTED** [review]
Sorry this took forever!

With regard to how graphs/maps should reset - By memory, on the V2 dashboard it reset when either the page was reset (not ideal) or when a `start` message was received.

I think the ideal case would be it only resets when we get a `start` message, and your local storage stuff has done a great job of fixing this. The only problem I can see is that if the `start` message is sent while the tab is closed, the graphs will never be reset. Any ideas how we could mitigate this? Things are complicated by the fact that the DAS doesn't have an RTC for knowing the timestamp of when the last reset was.

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (kdav108) [reply client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx:62]: This document hasn't been reformatted with `prettier` is what I meant

**Comment by haotongwang** [comment]
This PR uses `sessionStorage` which is reset when the tab is closed, but the data will persist when we refresh the page or switch to another tab.

File: client/.vscode/settings.json
Comment (haotongwang) [reply client/.vscode/settings.json]: I'm using different workspaces for `client` and `server` (pretty much treating them as different projects).

I thought `.vscode` was already in the `.gitignore` so didn't mean to commit this. Have removed it in 0b172aa

File: client/.eslintrc
Comment (haotongwang) [reply client/.eslintrc]: Yes, because we have many files that are indented with 4. I've changed it to a warning so storybook should build now, though with many warnings.
11b4ae3

File: client/src/components/common/charts/CadenceTimeChart.js
Comment (haotongwang) [reply client/src/components/common/charts/CadenceTimeChart.js:2]: The files are touched due to moving `ScatterChart` into `common` so unfortunately, even after sorting there are still many files in this PR.
7966df2

File: client/src/components/common/charts/LocationMap.tsx
Comment (haotongwang) [reply client/src/components/common/charts/LocationMap.tsx]: done 
2790c1b

File: client/src/components/v3/dashboard/V3LocationMap.tsx
Comment (haotongwang) [reply client/src/components/v3/dashboard/V3LocationMap.tsx]: done
3fb9392

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (haotongwang) [reply client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx]: Good point on keeping track of maximum! As someone doing FIT3155 and having done FIT2004, you would think this would be ingrained in me already.
7cf0d35

**Comment by haotongwang** [comment]
So it turns out making it mobile friendly wasn't too big of a task, so I've added it to this PR (sorry it's gotten a bit big, but most of it is css files).

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (hallgchris) [inline client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx]: Just realised this now - we're actually putting two datapoints on the chart for every update, once when the speed changes, and again when the distance changes. If we remove distances from the dependency array (there might be a better way to do this though?) we get a much smoother looking graph:
image

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (hallgchris) [inline client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx]: Another thing I only just realised, and I apologise - Can we use GPS speed instead of reed here? At high speeds, GPS speed is more accurate as there's no uncertainty in wheel radius (which is a function of the tyre we used, the tyre pressure, etc)

Ideally we'd use GPS but then fall back to reed if GPS is unavailable, but don't worry about this for now. Something to worry about later :)

Also, unsure what the name `point` means - `speed` or `velocity`?

**Review by hallgchris — APPROVED** [review]
Looking much nicer on mobile! We can do some tweaking later but it's usable now which is fantastic.

Just a couple quick changes and this is looking good! Happy to merge when these are addressed and tested.

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (haotongwang) [reply client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx]: `point` is meant to be a point on the graph and I named it that way because I didn't know whether `ReedVelocity` was meant to be speed or velocity.
Changed to GPS speed and renamed `point` to `speed`.
ed02064

File: client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx
Comment (haotongwang) [reply client/src/components/v3/dashboard/V3SpeedDistanceChart.tsx]: The only thing would be that if we had the same speed, the graph wouldn't update - though that's probably (?) not a very common occurrence.
Removed `distance` dependency.
299a0b5

**Comment by haotongwang** [comment]
Dismissing @kdav108 review due to it being quite old and did not involve any complex changes. The changes requested have been addressed.
I'll dismiss and merge tomorrow to give time for objections / further reviews.

**Pull request merged**
